### PR TITLE
	storage: fix incorrect LookupRange error causing TestSplitSnapshotRace_SplitWins flakiness

### LINKED ISSUE #8868: Failed tests (21860): TestSplitSnapshotRace_SplitWins
https://circleci.com/gh/cockroachdb/cockroach/21860

```
I160827 11:25:20.251358 util/log/trace.go:64  reply error Increment ["a",/Min): key range "a"-"a\x00" could not be located within a range on store
W160827 11:25:20.251571 kv/dist_sender.go:816  key range "a"-"a\x00" could not be located within a range on store
I160827 11:25:20.253728 kv/dist_sender.go:1035  application error: key range "a"-"a\x00" could not be located within a range on store
I160827 11:25:20.253868 util/log/trace.go:64  reply error Increment ["a",/Min): key range "a"-"a\x00" could not be located within a range on store
W160827 11:25:20.254101 kv/dist_sender.go:816  key range "a"-"a\x00" could not be located within a range on store
panic: test timed out after 5m0s

```

### PR DESCRIPTION
**Pull request by RaduBerinde**
The test occasionally times out under `stressrace`. The culprit is a condition
where the first replica doesn't find the range and incorrectly returns
RangeKeyMismatchError. The request is not retried on other replicas, but is then
retried endlessly by dist_sender.

The fix is to return a `RangeNotFound` in this case. Rewriting the code in the
function a bit to make it more readable.

Fixes #8868.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/8988)

### CODE REVIEW

File: storage/stores.go [hunk 1]
```diff
@@ -192,44 +192,51 @@ func (ls *Stores) Send(ctx context.Context, ba roachpb.BatchRequest) (*roachpb.B
...
+				start, end, desc.StartKey, desc.EndKey)
+			err := roachpb.NewRangeKeyMismatchError(start.AsRawKey(), end.AsRawKey(), desc)
+			return 0, roachpb.ReplicaDescriptor{}, err
+		}
+
+		var err error
+		repDesc, err = replica.GetReplicaDescriptor()
+		if err != nil {
+			if _, ok := err.(*roachpb.RangeNotFoundError); ok {
+				// Ignore this replica.
```
Comment (tschottdorf) [inline storage/stores.go]: More words

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/client_test.go, line 398 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988#-KQWfBQJQ2f9xlZ54iWY:-KQWfBQKxY_7kOGWHo_g:badp6l6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9b4f1b32a8850cff33e1750f38a0336083a0616/storage/client_test.go#L398)):*
> ```Go
> 	nodeIndex := int(rep.NodeID) - 1
> 	if log.V(2) {
> 		log.Infof(context.TODO(), "SendNext nodeIndex=%d", nodeIndex)
> ```

this is test code - feel free to remove the V(2)

---

*[storage/stores.go, line 234 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988#-KQWeihQazcQrMuvDLuh:-KQWeihQazcQrMuvDLui:b9tftdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9b4f1b32a8850cff33e1750f38a0336083a0616/storage/stores.go#L234)):*
> ```Go
> 
> 		repDescFound = true
> 		rangeID = replica.RangeID
> ```

can you move this up to where `repDesc` is assigned? slightly more expected there.

```
rangeID = replica.RangeID

var err error
repDesc, err = replica.GetReplicaDescriptor()
...
```

---

*[storage/stores.go, line 237 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988#-KQWf4tklvjo8CztImNl:-KQWf4tklvjo8CztImNm:bocwtj2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9b4f1b32a8850cff33e1750f38a0336083a0616/storage/stores.go#L237)):*
> ```Go
> 	}
> 	if !repDescFound {
> 		return 0, roachpb.ReplicaDescriptor{}, roachpb.NewRangeNotFoundError(0)
> ```

This looks good to me. Can you write a focused test for it?

---

*[storage/stores_test.go, line 205 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988#-KQWdesiy-Vhp-qjzKtv:-KQWdesiy-Vhp-qjzKtw:b-fv1xxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/720b6d0dbe9292ee7e2034396b09a7177d8e41df/storage/stores_test.go#L205)):*
> ```Go
> 		if tc.expError != "" {
> 			if !testutils.IsError(err, tc.expError) {
> 				t.Errorf("got unexpected error %v", err)
> ```

you need to log the expected error here - it is no longer evident from the code. also, every one of these calls to `Errorf` should begin with the loop index variable "%d: "

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988#-:-KQWfEDi8hGtZDMAhXlC:bnfp4nl)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[storage/stores.go, line 221 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988#-KQWbACtfZw3Om_uNW-s-r2-221:-KQWgdbdHFhcS_0RF0cu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9b4f1b32a8850cff33e1750f38a0336083a0616/storage/stores.go#L221)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> More words
</details>
Done.

---

*[storage/stores.go, line 234 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988#-KQWeihQazcQrMuvDLuh:-KQWgcUOtHqqwU63K50h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9b4f1b32a8850cff33e1750f38a0336083a0616/storage/stores.go#L234)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you move this up to where `repDesc` is assigned? slightly more expected there.
> 
> ```
> rangeID = replica.RangeID
> 
> var err error
> repDesc, err = replica.GetReplicaDescriptor()
> ...
> ```
</details>
Done.

---

*[storage/stores.go, line 237 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988#-KQWf4tklvjo8CztImNl:-KQWjntn3UXF8zmnD7n3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9b4f1b32a8850cff33e1750f38a0336083a0616/storage/stores.go#L237)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> This looks good to me. Can you write a focused test for it?
</details>
Done.

---

*[storage/stores_test.go, line 205 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988#-KQWdesiy-Vhp-qjzKtv:-KQWiiQcqOrUrV5i1z3t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/720b6d0dbe9292ee7e2034396b09a7177d8e41df/storage/stores_test.go#L205)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> you need to log the expected error here - it is no longer evident from the code. also, every one of these calls to `Errorf` should begin with the loop index variable "%d: "
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988)*

**Comment by tamird** [comment]
Reviewed 3 of 3 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/stores.go, line 223 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988#-KQWkw0eG410dRC-1o_A:-KQWkw0eG410dRC-1o_B:byynlxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/284396abf5876189a773077c6eb83559aebf2c00/storage/stores.go#L223)):*
> ```Go
> 		if err != nil {
> 			if _, ok := err.(*roachpb.RangeNotFoundError); ok {
> 				// We are not holding a lock across this block so the range could have
> ```

"range could have gone away" isn't very descriptive. specifically, this would only happen if this replica was removed from the range via down-replication.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988)*

**Comment by tschottdorf** [comment]
Reviewed 3 of 3 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988)*

**Comment by RaduBerinde** [comment]
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/stores.go, line 223 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988#-KQWkw0eG410dRC-1o_A:-KQWn0aO-OBsKhFjSc00:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/284396abf5876189a773077c6eb83559aebf2c00/storage/stores.go#L223)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "range could have gone away" isn't very descriptive. specifically, this would only happen if this replica was removed from the range via down-replication.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988)*

**Comment by tamird** [comment]
Reviewed 2 of 3 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8988)*

**Pull request merged**